Update aSHM heatmap functionality

lkhilton opened the pull request:
This PR includes a new helper function to consistently handle genomic region arguments to various functions. aSHM heatmaps will no longer be generated by `get_mutation_frequency_bin_matrix` - instead this function now returns only a matrix of mutation frequencies per sliding window, and a new function `heatmap_mutation_frequency_bin` generates the heatmap. This enables users to quickly generate mutation frequency matrixes without needing to ensure metadata are available and tidy. 

# Pull Request Checklists

**Important:** When opening a pull request, keep only the applicable checklist and delete all other sections.

## Checklist for all PRs

### Required

- [x] I tested the new code for my use case (please provide a reproducible example of how you tested the new functionality)
```
library(GAMBLR)
library(tidyverse)

gambl_meta <- get_gambl_metadata(seq_type_filter = c("capture", "genome")) %>%
    filter(pathology == "DLBCL")

process_regions(only_regions = "MYC")

calc_mutation_frequency_sliding_windows(
    this_region = "8:128748352-128749427", 
    these_samples_metadata = gambl_meta
)

get_mutation_frequency_bin_matrix(
    these_samples_metadata = gambl_meta,
    only_regions = c("MYC", "BCL2", "BCL6")
)

heatmap_mutation_frequency_bin(
    these_samples_metadata = gambl_meta,
    only_regions = c("MYC", "BCL2", "BCL6"),
    sortByColumns = "lymphgen"
)
```

- [x] I ensured all dplyr functions that commonly conflict with other packages are fully qualified. 

This can be checked and addressed by running `check_functions.pl` and responding to the prompts. Test your code _after_ you do this.

NOTE FROM LAURA: I ran this script and it's cool but also clearly lots of people have not been running it because it had TONS of suggested changes unrelated to the changes I made today. 

- [x] I generated the documentation and checked for errors relating to the new function (e.g. `devtools::document()`) and added `NAMESPACE` and all other modified files in the root directory and under `man`. 

- [ ] I have rebuilt the site with `pkgdown::build_site(lazy = TRUE)` to reflect any updated package documentation.
This failed for me with the following error: 
```
Writing '404.html'
-- Building function reference -------------------------------------------------
Error : In '_pkgdown.yml', topic must be a known topic name or alias
✖ Not 'grch37_ashm_regions'
Error: callr subprocess failed: In '_pkgdown.yml', topic must be a known topic name or alias
✖ Not 'grch37_ashm_regions'
Type .Last.error.trace to see where the error occurred
```

### Optional but preferred with PRs

- [x] I updated and/or successfully knitted a vignette that relies on the modified code (which ones?)
ssm_tutorial.Rmd has been updated to use `heatmap_mutation_frequency_bin`. 

## Checklist for New Functions

### Required

- [x] I documented my function using [ROxygen style](https://jozef.io/r102-addin-roxytags/#:~:text=Inserting%20a%20skeleton%20%2D%20Do%20this,Shift%2BAlt%2BR%20).)

- [x] Adequate function documentation (see [new-function documentation template](https://github.com/morinlab/GAMBLR#title) for more info)

- [x] My function uses a library that isn't already a dependency of GAMBLR and I made the package aware of this dependency using the function documentation `import` statement. 

Example:
```
#' @return nothing
#' @export
#' @import tidyverse ggrepel
```

## Checklist for changes to existing code

- [x] I added/removed arguments to a function and updated documentation for all changed/new arguments

- [x] I tested the new code for compatibility with existing functionality in the Master branch (please provide a reprex of how you tested the original functionality)
This PR reflects a substantial enough change that backward compatibility is not maintained.

rdmorin commented on R/utilities.R:
```diff
@@ -805,181 +805,414 @@ trim_scale_expression = function(x){
+#' @param regions Character vector of genomic regions. If neither regions nor regions_df is specified, will use GAMBLR aSHM regions
+#' @param regions_df Data frame of genomic regions with column names "chrom", "start", "end", "name"
+#' @param region_padding Amount to pad the start and end coordinates by. [0]
+#' @param skip_regions Character vector of genes to drop from GAMBLR aSHM regions.
```
the skip and keep parameters seem specific to the aSHM region list. I think that's fine but we should probably be explicit in documenting that for the user. regions_df should also be documented as optional and we should specify that the default is ashm_regions for the specified projection.

lkhilton replied:
```If neither regions nor regions_df is specified, will use GAMBLR aSHM regions```
Is that clear that these are optional and what the default is? 
```Character vector of genes to drop from GAMBLR aSHM regions.```
How would you change this to make it clearer that this applies to the default?

rdmorin replied:
Oh I didn't see that, probably because I was focusing on regions_df. Maybe just duplicate that comment on the regions_df line?

rdmorin commented on R/utilities.R:
```diff
@@ -805,181 +805,414 @@ trim_scale_expression = function(x){
+#' @noRd
+#'
+#' @examples
+#' trimmed = trim_scale_expression(2)
```
Even though it's an internal function, maybe add a real example?

rdmorin commented on R/utilities.R:
```diff
@@ -805,181 +805,414 @@ trim_scale_expression = function(x){
+                            region_padding = 0,
+                            skip_regions = NULL,
+                            only_regions = NULL,
+                            projection = "grch37") {
```
What about an optional "with_chr_prefix" argument? The default could be to strip it (if necessary) and return chromosome names without prefix but if this is enabled, it would ensure they are prefixed (add if necessary).

rdmorin commented on R/utilities.R:
```diff
@@ -805,181 +805,414 @@ trim_scale_expression = function(x){
+                                          end_pos,
+                                          these_samples_metadata,
+                                          these_sample_ids,
+                                          maf = NULL,
```
Since some functions work with both, I try to use maf_df for maf data frame/object and maf_file for a path to a file.

rdmorin commented on R/utilities.R:
```diff
@@ -805,181 +805,414 @@ trim_scale_expression = function(x){
+
+  # Check for provided metadata, else use GAMBL metadata
+  if (missing(these_samples_metadata)) {
+    metadata <- get_gambl_metadata(
```
It might be risky to allow this function to work on samples of a different seq_type on one call to the function. If we make the user specify it at least they have to intentionally merge the output after, and hopefully also think about how it's likely not a good idea to do this in many scenarios.

rdmorin replied:
Another consideration here is those sample_id that have >1 seq type in GAMBL. Will this cause a problem in this function?

rdmorin replied:
So you think it's a good thing to allow users to run this function on a mix of exome data and genome data (possibly without realizing it)? That would be the default behaviour right now, wouldn't it?

lkhilton replied:
I addressed this by adding an argument to allow users to subset to one seq_type, but the default will remain the combination of capture and genome, utilizing the default seq_type_priority for get_gambl_metadata. 

Users should be encouraged to always provide metadata and not a vector of sample IDs so they can be confident that they're getting results only from the actual samples they want (I added a note about this to the id_ease documentation).

rdmorin commented on R/utilities.R:
```diff
@@ -805,181 +805,414 @@ trim_scale_expression = function(x){
+  } else {
+    these_sample_ids <- metadata$sample_id
+  }
+
```
How do we deal with values in these_sample_ids that are not in the metadata? Should we catch those and warn the user?

rdmorin commented on R/utilities.R:
```diff
@@ -805,181 +805,414 @@ trim_scale_expression = function(x){
+    if (is.null(regions_df)) {
+      message("Using default GAMBLR aSHM regions. ")
+      if (projection == "grch37") {
+        regions_df <- grch37_ashm_regions %>%
```
I see what you did there.

rdmorin commented on R/utilities.R:
```diff
@@ -805,181 +805,414 @@ trim_scale_expression = function(x){
+      regions_df$gene <- names(regions)
+    } else {
+      message("WARNING: Regions provided as an unnamed character vector. It is strongly recommended to provide a named vector with a unique name per region, or a bed file-formatted data frame. ")
+      regions_df$name <- regions_df$chromosome
```
What about defaulting to the name being paste0(chromosome,"_",start)?

lkhilton replied:
Good idea

rdmorin commented on R/viz.R:
```diff
@@ -658,180 +810,341 @@ pretty_lollipop_plot = function(maf_df,
+#'    regions_df = some_regions
+#' )
+#'
+
```
Renaming the function is sensible but might break other code. I wonder if there's a common strategy to handle this (e.g. by leaving a skeleton function with the old name that just calls the new function?

lkhilton replied:
The name of the old function is still in use though, it's job is just to actually return a matrix, not a heatmap. 

This is where I wish we had working package versioning and more people using renv for maintaining old functionality...

rdmorin commented on R/viz.R:
```diff
@@ -658,180 +810,341 @@ pretty_lollipop_plot = function(maf_df,
-      regions_df = grch37_ashm_regions #drop MYC and BCL2
-      regions_df = grch37_ashm_regions %>%
-        dplyr::filter(!gene %in% skip_regions)
+#' mut_count_matrix <- get_mutation_frequency_bin_matrix(
```
Example is out of date given the new function name

lkhilton replied:
Good catch

rdmorin commented on R/viz.R:
```diff
@@ -658,180 +810,341 @@ pretty_lollipop_plot = function(maf_df,
 #' #get ashm regions
 #' some_regions = grch37_ashm_regions
```
I suspect this example is out of date. Do we need to call the new function to get the regions in the right format?

rdmorin commented on vignettes/ssm_tutorial.Rmd:
```diff
@@ -343,7 +343,7 @@ dlbcl_bl_meta = collate_results(join_with_full_metadata = TRUE,
 
 #get ashm regions (grch37) and filter out a set of genes.
-some_regions = grch37_ashm_regions %>% 
+some_regions = GAMBLR.data::somatic_hypermutation_locations_GRCh37_v_latest %>% 
```
This can be reverted, right?

lkhilton replied:
Yep, will fix

rdmorin reviewed (changes requested):
Great addition! A few minor changes and some questions that may or may not lead to more changes.

lkhilton commented on R/utilities.R:
```diff
@@ -805,181 +805,414 @@ trim_scale_expression = function(x){
+    }
+  }
+
+  # Collapse regions with duplicate names
```
Yes, hence the warning that users should heed, but it would be resolved by your suggestion for how to name the regions vector by default.

Kdreval commented:
@lkhilton Please let me know if this is ready to merge or there is something to be done to address Ryan's comments? There seems to be some merge conflicts - do you want me to fix those?

lkhilton commented:
@rdmorin I think I've addressed all your comments and harmonized all the arguments to match the `get_ssm_by_region(s)` functions. 

@Kdreval and I spoke about separating these out into new scripts because he is working on implementing this across the package. Kostia, would you mind taking a look at this PR as well?

lkhilton commented:
Also if you want to do any testing, this code should all work: 
```

gambl_meta <- get_gambl_metadata(seq_type_filter = c("capture", "genome")) %>%
    filter(pathology == "DLBCL")

regions <- GAMBLR:::process_regions(only_regions = c("MYC", "BCL2", "BCL6"))

calc_mutation_frequency_bin_region(
    region = regions$regions_list[1], 
    these_samples_metadata = gambl_meta, 
) 

calc_mutation_frequency_bin_regions(
    these_samples_metadata = gambl_meta,
    regions_list = regions$regions_list[c(1, 7, 9)]
)

heatmap_mutation_frequency_bin(
    these_samples_metadata = gambl_meta,
    regions_list = regions$regions_list[c(1, 7, 9)],
    sortByColumns = c("lymphgen", "seq_type")
)
```

Kdreval commented:
Thanks so much for giving the test code!
It returned some errors for me though. Here is the error message:
```
r$> calc_mutation_frequency_bin_regions(
        these_samples_metadata = gambl_meta,
        regions_list = regions$regions_list[c(1, 7, 9)]
    )
Warning: Multiple regions in the provided data frame have the same name. Merging these entries based on min(start) and max(end) per name value. 
processing bins of size 500 across 8106 bp region
Using GAMBLR::get_ssm_by_region...
                                                                                                                                                                                                    Joining with `by = join_by(sample_id)`
Joining with `by = join_by(window_start)`
Error in `dplyr::bind_rows()`:
! Argument 2 must be a data frame or a named atomic vector.
Run `rlang::last_trace()` to see where the error occurred.
Warning message:
In parallel::mclapply(names(regions), function(x) { :
  scheduled core 2 encountered error in user code, all values of the job will be affected

r$> heatmap_mutation_frequency_bin(
        these_samples_metadata = gambl_meta,
        regions_list = regions$regions_list[c(1, 7, 9)],
        sortByColumns = c("lymphgen", "seq_type")
    )
Warning: Multiple regions in the provided data frame have the same name. Merging these entries based on min(start) and max(end) per name value. 
processing bins of size 500 across 8106 bp region
Using GAMBLR::get_ssm_by_region...
                                                                                                                                                                                                    Joining with `by = join_by(sample_id)`
Joining with `by = join_by(window_start)`
Error in `dplyr::bind_rows()`:
! Argument 2 must be a data frame or a named atomic vector.
Run `rlang::last_trace()` to see where the error occurred.
Warning message:
In parallel::mclapply(names(regions), function(x) { :
  scheduled core 2 encountered error in user code, all values of the job will be affected
```

Probably related to NAs here in the `regions_list`?
```
r$> regions$regions_list[c(1, 7, 9)]
               BCL6_TSS                    <NA>                    <NA> 
"3:187458526-187464632"                      NA                      NA 
```

Here is the object `regions`:
```
r$> regions
$regions_list
               BCL6_TSS                 MYC_TSS           BCL2_intronic                BCL2_TSS 
"3:187458526-187464632" "8:128748352-128749427"  "18:60796984-60814103"  "18:60982728-60988342" 

$regions_bed
  chrom     start       end gene   region regulatory_comment          name
1     3 187458526 187464632 BCL6      TSS               <NA>      BCL6_TSS
2     8 128748352 128749427  MYC      TSS    active_promoter       MYC_TSS
3    18  60796984  60814103 BCL2 intronic    strong_enhancer BCL2_intronic
4    18  60982728  60988342 BCL2      TSS    active_promoter      BCL2_TSS
```
If i change the regions_list to have something from the bed file and no NAs like this: `regions$regions_list[c(1, 2, 4)]`, the code works and the last command generates this plot
![image](https://github.com/morinlab/GAMBLR/assets/47579823/90a45326-70b8-4c23-a116-598de2dfdf9c)

So I think it is working as expected, thank you for updating the code!

The GitHub actions seem to fail on the latest commit for documentation reasons. [This](https://github.com/morinlab/GAMBLR/actions/runs/5581140891/jobs/10198966941#step:6:5008) is the relevant part from the log, I think it just needs a documentation update and the PR will be ready to merge :rocket:

lkhilton commented:
Thanks for testing it out, Kostia. Were you using the latest GAMBLR.data? I think my testing used the latest which seems to have a bunch more BCL6 regions than it used to, so subsetting the named vector returned by process_regions by position probably looked at empty slots if you were using older GAMBLR.data?

Kdreval commented:
Thanks Laura, this was exactly the issue - I updated to the latest version and example code worked right away with no issues!
